### PR TITLE
Fixes SI-7943 -- make `TrieMap.getOrElseUpdate` atomic.

### PR DESCRIPTION
**Pull request by axel22**
Add a `computeIfAbsent` method to `TrieMap`.
The signature and contract of this method corresponds closely to the
`computeIfAbsent` from `java.util.concurrent.ConcurrentMap`.
Eventually, this method should be added to
`scala.collection.concurrent.Map`.

Override `getOrElseUpdate` to call `computeIfAbsent`.

Add tests for both methods.

Please let me know if this should go to 2.12.x.

Review by @retronym

### CODE REVIEW

**Comment by axel22** [comment]
I'm guessing that this causes binary incompatibilities, so we should target 2.12.x?
(or make the new `computeIfAbsent` method private for now)

Also, review by @Ichoran

File: src/library/scala/collection/concurrent/TrieMap.scala [hunk 1]
```diff
@@ -873,6 +873,42 @@ extends scala.collection.concurrent.Map[K, V]
...
+        val hc = computeHash(k)
+        insertifhc(k, hc, v, INode.KEY_ABSENT) match {
+          case Some(oldv) => oldv
+          case None => v
+        }
+      }
+    }
+  }
+
+  override def getOrElseUpdate(k: K, v: =>V): V = computeIfAbsent(k, _ => v)
```
Comment (Ichoran) [inline src/library/scala/collection/concurrent/TrieMap.scala]: `v` is called `op` in the parent.  Should keep the names the same (even if `op` is a strange name--maybe we'll change everything to `default` someday?).

File: src/library/scala/collection/concurrent/TrieMap.scala [hunk 2]
```diff
@@ -873,6 +873,42 @@ extends scala.collection.concurrent.Map[K, V]
...
+   *  that exception is rethrown.
+   *  
+   *  Note: this method may invoke the specified function `f` more than once,
+   *  even if the key is in the end not added to the map.
+   *
+   *  @param k      the key to modify
+   *  @param f      the mapping function that computes the value from the key
+   *  @return       the newly added value
+   */
+  def computeIfAbsent(k: K, f: K => V): V = {
```
Comment (Ichoran) [inline src/library/scala/collection/concurrent/TrieMap.scala]: The name may be misleading.  It doesn't just compute; it inserts, also.  However, I wonder why we need this method at all?  The caller has `k` in hand already, and we're not going to use `f` on any other value than `k`, so why not let them do the computation?  (That is, what's wrong with `getOrElseUpdate`, aside from the overly long name?)

**Comment by Ichoran** [comment]
Yes, this will have to go in 2.12 because of binary compatibility.  (But I would suggest just overriding getOrElseUpdate rather than deferring to a new method; the use cases for the two methods seem identical.)

File: src/library/scala/collection/concurrent/TrieMap.scala (hunk 1 above)
Comment (axel22) [inline src/library/scala/collection/concurrent/TrieMap.scala]: Good point - will change that.

File: src/library/scala/collection/concurrent/TrieMap.scala (hunk 2 above)
Comment (axel22) [inline src/library/scala/collection/concurrent/TrieMap.scala]: I've chosen this name and signature mainly to be consistent with the `computeIfAbsent` in `java.util.concurrent`, the way we've done it for pre-JDK8 for other `ConcurrentMap` methods. My guess is that eventually we want to do this, to ease the conversions between Java and Scala concurent maps.

I see your concern, however. I could just keep `getOrElseUpdate` in this PR.

**Comment by axel22** [comment]
Sure - will resubmit.

**Pull request closed without merging**